roachtest: add Sqlapp consistency check tests

nvb opened the pull request:
Fixes cockroachlabs/rubrik#19.

This change adds a series of roachtests that run the Sqlapp consistency
check suite. This suite consists of three applications:
- distributed_semaphore
- filesystem_simulator
- jobworker/jobcoordinator

Each application was designed to be used to check the consistency of a
database under load. They each launch a set of workers who perform database
operations while another worker periodically checks invariants to capture
any inconsistencies. The apps have been pulled from
https://github.com/scaledata/rksql/tree/master/src/go/src/rubrik/sqlapp.

I have a script similar to `loadgen/teamcity-push.sh` that compiles and
uploads all binaries from https://github.com/scaledata/rksql into s3 under
the `cockroach` bucket and the `rubrik` folder. I'm not sure where that
should live.

This PR works on pre-built binaries that include the changes proposed in
https://github.com/scaledata/rksql/pull/1.

Release note: None

nvb committed:
roachtest: add Sqlapp consistency check tests

This change adds a series of roachtests that run the Sqlapp consistency
check suite. This suite consists of three applications:
- distributed_semaphore
- filesystem_simulator
- jobworker/jobcoordinator

Each application was designed to be used to check the consistency of a
database under load. They each launch a set of workers who perform database
operations while another worker periodically checks invariants to capture
any inconsistencies. The apps have been pulled from
github.com/scaledata/rksql/tree/master/src/go/src/rubrik/sqlapp.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24189)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/cmd/roachtest/sqlapp.go, line 72 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24189#-L8OxLoyVOpFCSziMK4s:-L8OxLoz2tEm75ssC_O0:b6xbkwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17cb3dae548bf44b0dc9affcc1e691c3e4619e8/pkg/cmd/roachtest/sqlapp.go#L72)):*
> ```Go
> 		Binary:    app,
> 		// TODO DURING REVIEW: We could use LATEST, but I suspect reproducibility
> 		// is more important than automatically picking up changes.
> ```

If we do use LATEST, we'd want to display the sha in the test log for reproducibility of failures. My instinct is we should be using LATEST.

---

*[pkg/cmd/roachtest/sqlapp.go, line 96 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24189#-L8OxebzzfTQDhGa2eu5:-L8OxebzzfTQDhGa2eu6:bvl8fxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17cb3dae548bf44b0dc9affcc1e691c3e4619e8/pkg/cmd/roachtest/sqlapp.go#L96)):*
> ```Go
> 		addrs = append(addrs, url.Host)
> 	}
> 	addrStr := strings.Join(addrs, ",")
> ```

Heh, @bdarnell is dealing with this exact same problem in #24187. I think `roachprod` or `cluster` need to provide an easy means of getting the internal IP address for a node.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24189#-:-L8OyLIsZF6I-EHcRS2C:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/cmd/roachtest/sqlapp.go, line 72 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24189#-L8OxLoyVOpFCSziMK4s:-L8Q3rUdDQ-A4wcdyQ7T:b-3xmai) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17cb3dae548bf44b0dc9affcc1e691c3e4619e8/pkg/cmd/roachtest/sqlapp.go#L72)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

If we do use LATEST, we'd want to display the sha in the test log for reproducibility of failures. My instinct is we should be using LATEST.
</blockquote></details>

What @petermattis said.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24189#-:-L8Q4GIK16FUR2HL1xe_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
"Sqlapp" is awfully generic; this needs a better name.

tschottdorf commented:
Instead of renaming `sqlapp` upstream, it might be sufficient to prefix everything with `rubrik_`. The `sqlapp` binary is already under the `rubrik` component, so that would be consistent.

nvanbenschoten commented:
Per the conversation earlier today, I prefixed the file and the test name with `scaledata`.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/sqlapp.go, line 72 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24189#-L8OxLoyVOpFCSziMK4s:-L8YUr_8zSh3eYlML-hZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17cb3dae548bf44b0dc9affcc1e691c3e4619e8/pkg/cmd/roachtest/sqlapp.go#L72)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

What @petermattis said.
</blockquote></details>

Done.

---

*[pkg/cmd/roachtest/sqlapp.go, line 96 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24189#-L8OxebzzfTQDhGa2eu5:-L8YXp8zdL9P0KZxYXH6:b4ftrlf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17cb3dae548bf44b0dc9affcc1e691c3e4619e8/pkg/cmd/roachtest/sqlapp.go#L96)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Heh, @bdarnell is dealing with this exact same problem in #24187. I think `roachprod` or `cluster` need to provide an easy means of getting the internal IP address for a node.
</blockquote></details>

Should we just change `cluster.PGUrl` to return the internal IP instead of the external IP? I expect that in these tests any use of this IP will be from another GCE node.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24189#-:-L8YUzkes14UX9r6-83m:babrkmp)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/roachtest/scaledata.go, line 72 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24189#-L8OxLoyVOpFCSziMK4s:-L8YizQvxLIgj8ZXsfCv:bxoy0wo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17cb3dae548bf44b0dc9affcc1e691c3e4619e8/pkg/cmd/roachtest/scaledata.go#L72)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Does `binfetcher.Download` display the SHA?

---

*[pkg/cmd/roachtest/scaledata.go, line 96 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24189#-L8OxebzzfTQDhGa2eu5:-L8YhN19wZYjpb2mwqT4:b-a4nbiz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17cb3dae548bf44b0dc9affcc1e691c3e4619e8/pkg/cmd/roachtest/scaledata.go#L96)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we just change `cluster.PGUrl` to return the internal IP instead of the external IP? I expect that in these tests any use of this IP will be from another GCE node.
</blockquote></details>

The usage here and in `decommission.go` wants the internal IP address. The usage from `cluster.Conn` wants the external IP address. Perhaps we need both `ExternalPGUrl` and `InternalPGUrl`. Or perhaps the existing `cluster.PGUrl` should return internal URLs and a new `cluster.ExternalPGUrl` should be provided for use by `cluster.Conn`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24189)*
<!-- Sent from Reviewable.io -->

nvb committed:
binfetcher: log redirect url for LATEST binaries

Release note: None

nvb committed:
roachtest: make handling of internal and external IPs more explicit

This change adds the following methods to `cluster` and cleans up
some of the roactests by using them:
- `InternalPGUrl`
- `ExternalPGUrl`
- `InternalIP`
- `ExternalIP`

Release note: None

nvanbenschoten commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/cmd/roachtest/scaledata.go, line 72 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24189#-L8OxLoyVOpFCSziMK4s:-L8Yp-dNrnHU-kV5OQdp:b6u4qos) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17cb3dae548bf44b0dc9affcc1e691c3e4619e8/pkg/cmd/roachtest/scaledata.go#L72)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Does `binfetcher.Download` display the SHA?
</blockquote></details>

Whoops, I missed that. It does now. 

---

*[pkg/cmd/roachtest/scaledata.go, line 96 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24189#-L8OxebzzfTQDhGa2eu5:-L8Z2UgWd-HUC-qKz1wg:b-dsz9k1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17cb3dae548bf44b0dc9affcc1e691c3e4619e8/pkg/cmd/roachtest/scaledata.go#L96)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The usage here and in `decommission.go` wants the internal IP address. The usage from `cluster.Conn` wants the external IP address. Perhaps we need both `ExternalPGUrl` and `InternalPGUrl`. Or perhaps the existing `cluster.PGUrl` should return internal URLs and a new `cluster.ExternalPGUrl` should be provided for use by `cluster.Conn`.
</blockquote></details>

Done. Depends on https://github.com/cockroachdb/roachprod/pull/135.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24189)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTRs!
